### PR TITLE
Move StructuredCommandLine redaction to redactor

### PR DESCRIPTION
**Pull request by bduffany**
* Redact env vars + remote headers and strip URL secrets from structured command line
* Parse allowed env vars from the started event instead of from the build metadata event (since build metadata can come later on)
  * Confirmed that we do see ALLOW_ENV in the started event's options description field, even if it is indirectly specified via a bazelrc config

---

**Version bump**: Patch

### CODE REVIEW

File: server/util/redact/redact_test.go [hunk 1]
```diff
@@ -47,44 +48,92 @@ func TestRedactMetadata_UnstructuredCommandLine_RemovesArgs(t *testing.T) {
...
+		CombinedForm: "--remote_cache_header=x-buildbuddy-api-key=bbapikey",
+		OptionName:   "remote_cache_header",
+		OptionValue:  "x-buildbuddy-api-key=bbapikey",
+	}
+	nonEnvURLOption := &clpb.Option{
+		CombinedForm: "--some_url=https://password@foo.com",
+		OptionName:   "some_url",
+		OptionValue:  "https://password@foo.com",
+	}
+	structuredCommandLine := &clpb.CommandLine{
```
Comment (vadimberezniker) [inline server/util/redact/redact_test.go:92]: Do we get any value out of passing all of the options at once, but making the assertions in a loop? It makes the test harder to read.

Could we do something like the following instead?
option = &clpb.Option{...}
redactor.RedactMetadata(structuredCommandLineEvent(option))
assert.Equal(t, ...)

You could still use a loop to go over the the inputs & expected output combinations if you wanted to.

**Review by vadimberezniker — APPROVED** [review]
LGTM

File: server/build_event_protocol/event_parser/event_parser.go [hunk 2]
```diff
@@ -10,78 +9,34 @@ import (
...
-							envVarMap[parts[0]] = parts[1]
-						}
-						if isAllowedEnvVar(parts[0], allowedEnvVars) {
-							continue
-						}
-
-						option.OptionValue = strings.Join([]string{parts[0], envVarRedactedPlaceholder}, envVarSeparator)
-						option.CombinedForm = envVarPrefix + envVarOptionName + envVarSeparator + parts[0] + envVarSeparator + envVarRedactedPlaceholder
+	for _, section := range commandLine.Sections {
+		if p, ok := section.SectionType.(*command_line.CommandLineSection_OptionList); ok {
```
[the remark this answers is not in the document]
Comment (bduffany) [reply server/build_event_protocol/event_parser/event_parser.go]: Done

File: server/util/redact/redact.go [hunk 3]
```diff
@@ -73,14 +88,113 @@ func stripRepoURLCredentialsFromWorkspaceStatus(status *bespb.WorkspaceStatus) {
...
+	}
+}
+
+func redactStructuredCommandLine(commandLine *clpb.CommandLine, allowedEnvVars []string) map[string]string {
+	envVarMap := make(map[string]string)
+	if commandLine == nil {
+		return envVarMap
+	}
+	for _, section := range commandLine.Sections {
+		if p, ok := section.SectionType.(*clpb.CommandLineSection_OptionList); ok {
```
[the remark this answers is not in the document]
Comment (bduffany) [reply server/util/redact/redact.go]: Done

File: server/util/redact/redact_test.go (hunk 1 above)
Comment (bduffany) [reply server/util/redact/redact_test.go]: Very much improved -- thanks for the suggestion!

**Pull request merged**